### DISCUSSION TITLE
Jetson kernel development

### DISCUSSION THREAD

**Discussion by Jesseeee** [OE4T/meta-tegra#1126] [General]
Hello,

I'm pretty new to the whole yocto thing but we're trying to develop something using the xavier/orin, and for this we need to add some of our own drivers and more generally just be able to develop on this kernel, but I'm a bit stuck on how to best go about this.
Currently we already build a tegra-demo image with some extras to get our application running and so far so good.
Would following these steps but changing linux-yocto to linux-tegra suffice?
https://docs.yoctoproject.org/kernel-dev/common.html#getting-ready-for-traditional-kernel-development
https://docs.yoctoproject.org/kernel-dev/common.html#using-traditional-kernel-development-to-patch-the-kernel
Or what would be the best way to go about modifying the jetson kernel or figuring out how it works? Because the whole patchwork of modified makefiles and overlays can be pretty confusing.

Thanks in advance,
Jesse

**Comment by madisongh** [comment]
Depending on how extensive your changes are going to be, it might be simpler to just fork the linux-tegra repo and create a branch with your modifications on it.  It's easy to add a bbappend file for the linux-tegra recipe to have it point to your repo, branch, and SRCREV.   That's what I do in my projects.

The Yocto documentation is more oriented towards the Yocto Project-hosted kernel, which combines sources from one repo with sets of config fragments and patches from another based on the target platform, which adds some complications to the process.  None of that is necessary here.

**Reply by Jesseeee** [reply]
When doing that, for easy local development, I presume I can just do something similar they describe in the docs e.g.
```
SRC_URI:pn-linux-yocto = "git:///path-to/linux-yocto-4.12;protocol=file;name=machine;branch=standard/base; \
                          git:///path-to/yocto-kernel-cache;protocol=file;type=kmeta;name=meta;branch=yocto-4.12;destsuffix=${KMETA}"
SRCREV_meta:qemux86 = "${AUTOREV}"
SRCREV_machine:qemux86 = "${AUTOREV}"
```
And just remove the kernel-cache/meta part and refer to linux-tegra and my own machine (currently xavier devkit), something like the following?
```
SRC_URI:pn-linux-tegra = "git:///path-to/linux-tegra-5.10;protocol=file;name=machine;branch=${MY_BRANCH};"
SRCREV_machine:jetson-agx-xavier-devkit = "${AUTOREV}"
```

And thanks a lot for the help!

**Reply by madisongh** [reply]
It's not even that complicated.  You just need to set `SRC_REPO:pn-linux-tegra` to point to your fork (see [this setting](https://github.com/OE4T/meta-tegra/blob/458d71ab8201b09ce76c5e25509c7ae755922cfc/recipes-kernel/linux/linux-tegra_5.10.bb#L25) in the kernel recipe), and `SRCREV:pn-linux-tegra`.  (and likewise for `SRCBRANCH` if you create your own branch). We don't use the Yocto kernel's "meta/machine" split.
